mlx/backend/metal/utils.h

```diff
@@ -162,6 +163,48 @@ collapse_contiguous_dims(Arrays... xs) {
       std::vector<array>{std::forward<Arrays>(xs)...});
 }
 
+#ifdef MLX_METAL_DEBUG
+
+NS::String* _debug_make_string(std::ostringstream& os) {
+  return NS::String::string(os.str().c_str(), NS::UTF8StringEncoding);
+}
+
+void _debug_set_stream_queue_label(MTL::CommandQueue* queue, int index) {
+  std::ostringstream label;
+  label << "Stream " << index;
+  queue->setLabel(_debug_make_string(label));
+}
+
+void _debug_push_group(Primitive& primitive) {
+  std::ostringstream group;
+  primitive.print(group);
+
+  auto stream = primitive.stream();
+  auto& device = metal::device(stream.device);
+  auto compute_encoder = device.get_command_encoder(stream.index);
+  compute_encoder->pushDebugGroup(_debug_make_string(group));
+}
+
+void _debug_pop_group(Primitive& primitive) {
+  auto stream = primitive.stream();
+  auto& device = metal::device(stream.device);
+  auto compute_encoder = device.get_command_encoder(stream.index);
+  compute_encoder->popDebugGroup();
+}
```

Comment:
Very interesting, thanks for the context. 

### Command Buffer

Push/pop on a command buffer won't group sub-encoder method calls:

<img width="320" alt="mlx-metal-capture-command-buffer-push-pop" src="https://github.com/ml-explore/mlx/assets/7120397/fcae69b7-352b-4dd3-9d63-a1e61ab00c0e">

### MPS Example

Below is a screenshot of the MPS `matmul` on the following workload:

```cpp
auto a = array({1.0f, 2.0f, 3.0f, 4.0f}, {2, 2});
auto x = matmul(a, a, s2);
std::cout << x << std::endl;
```

<img width="320" alt="mlx-metal-capture-mps" src="https://github.com/ml-explore/mlx/assets/7120397/6c204610-a2f4-4fe0-81d5-36d14e0664bf">

Besides the incorrect debug group boundaries (as expected), `debug_push_group` creates an encoder for the stream (via `get_command_encoder`), only to end encoding early. I will ensure whatever approach we land on will never create new command encoders.

---

### Alternatives

#### Per-Primitive Encoder

If each primitive's `eval_gpu` used a different compute encoder, the command buffer labeling would work. In fact, an improvement on this would be to simply label the encoder:

<img width="320" alt="mlx-metal-capture-encoder-labels" src="https://github.com/ml-explore/mlx/assets/7120397/7bbcb973-a647-46eb-89e6-837634724c48">

The obvious downside is excess command encoder allocations.

#### "Will Evaluate" Primitive Method

```cpp
// Called just before the primitive is evaluated. Returns a set of flags or
// an object detailing how the evaluation will happen. 
virtual uint64_t will_eval(const std::vector<array>& inputs, std::vector<array>& outputs) {
  return 0;
}
```

Some example cases where it could be overridden:

```
uint64_t Matmul::will_eval(...) {
  uint64_t flags = 0;

  if (will_use_mps(...)) {
    flags |= eval_flag_unused_stream_encoder;
  }

  return flags;
}
```

```
uint64_t Synchronizer::will_eval(...) {
  return eval_flag_noop;
}
```

Then based on the return value, the appropriate debug action could be taken (e.g., push group on command buffer instead of encoder).

Reply:
Could I trouble you to elaborate on what you meant by `Push/pop on a command buffer won't group sub-encoder method calls` ? I don't think I grasped it - does the frame capture not extend beyond one command encoder ?
Also, if needed, you can always go back to https://github.com/ml-explore/mlx/blob/main/mlx/backend/metal/metal.cpp and set max buffers per op = 1 when running in debug mode so that one primitive always has its own command buffer if that's the issue

Reply:
> Could I trouble you to elaborate on what you meant by Push/pop on a command buffer won't group sub-encoder method calls ?

In the "Command Buffer" example above, the second dispatch in the `Arange` group is actually the addition.

> I don't think I grasped it - does the frame capture not extend beyond one command encoder ?

The capture includes all command buffers.

---

> one primitive always has its own command buffer

I really like this simplification, thanks for pointing out `max_ops_per_buffer()`. With one command buffer per primitive, we don't need to push/pop groups. Instead, we can simply label the command buffer. Here's what that looks like:

<img width="320" alt="mlx-metal-capture-one-command-buffer-per-op" src="https://github.com/ml-explore/mlx/assets/7120397/d3cbf017-f568-42cc-9efd-274e1a0c60a6">

and in the MPS case:

<img width="320" alt="mlx-metal-capture-mps-labeled-command-buffer" src="https://github.com/ml-explore/mlx/assets/7120397/5a3dd232-6f02-4e97-85f6-b10eec3637ee">